### DISCUSSION TITLE
dual2nd grad -nan issue

### DISCUSSION THREAD

**Discussion by Manushanker** [autodiff/autodiff#300] [General]
HI Allan,

I am computing a Hessian using the dual2nd  forward type. The computed Jacobian and Hessian are all nan. When I debugged the code I found that for certain dual2nd  variable it has grad (under val) to be -nan(ind) and val and grad under grad is also -nan(ind). Can you please let me know what is the reason for nan? If need I can share you my code. 

Thanks,

Manushanker

**Comment by Manushanker** [comment]
I used the function 'isnan' to identify the dual2nd variables with val.grad, grad.val, grad.grad equal to nan and forced their value to be zero, then the library computes the correct Jacobian and Hessian values. I still would like to know why the val.grad becomes nan even though the value of the variable is not nan.

**Reply by allanleal** [reply]
This can happen if the function does not have derivative at the point you are evaluating it. Consider `f(x) = sqrt(x)`. `f(0) = 0` but `fx(0)` is undefined.

**Reply by Manushanker** [reply]
Thanks for a quick response. I will try to avoid cases where f(0)=0.

**Reply by allanleal** [reply]
Note `f(0)` is not always an issue. For example, `f(x) = exp(x)`. `f(0) = 1`, `fx(0) = 1`.

**Reply by Manushanker** [reply]
My function is not a single function, but a set of functions. For a certain part of the function the val.grad value becomes nan, which makes all the dependent variables also to be val.grad nan. The code is bit involved than simple sqrt(x) or exp(x), I will used your suggestion to narrow down the issue. Thanks a lot!

**Reply by allanleal** [reply]
Yes, that's clear. Somewhere inside your more complicated function, with multiple variables, you have a mathematical operation at a certain value for which derivative does not exist. Check this to get a runtime exception so that you can debug and pinpoint where the error is happening: https://stackoverflow.com/a/3616017/418875

**Reply by Manushanker** [reply]
Thanks for the stackoverflow link. I will keep you updated!

**Reply by Manushanker** [reply]
The reason for nan was the value of x in sqrt(x) was zero, which caused the derivative to be nan. Thanks for your time and help!
